picard/ui/statusindicator.py

```diff
@@ -30,33 +33,93 @@
+        if total_pending != last_pending:
+            previous_done_files = self._max_pending_files - self._last_pending_files
+            previous_done_files = previous_done_files if previous_done_files >= 0 else 0
+            previous_done_requests = self._max_pending_requests - self._last_pending_requests
+            previous_done_requests = previous_done_requests if previous_done_requests >= 0 else 0
+            self._max_pending_files = max(self._max_pending_files, previous_done_files + pending_files)
+            self._max_pending_requests = max(self._max_pending_requests, previous_done_requests + pending_requests)
+            self._last_pending_files = pending_files
+            self._last_pending_requests = pending_requests
+
+        if total_pending == 0 or (self._max_pending_files == 0 and self._max_pending_requests == 0):
+            self._max_pending_files = 0
+            self._max_pending_requests = 0
+            self._last_pending_files = 0
+            self._last_pending_requests = 0
+            self._prev_time = time.time()
             self.hide_progress()
+            # Turn of eta label if nothing is being processed
+            if self._eta_label:
+                self._eta_label.setVisible(False)
             return
 
-        progress = 1 - (total_pending / self._max_pending)
-        self.set_progress(progress)
+        if total_pending != last_pending:
+            current_time = time.time()
+
+            # time since we started processing this batch
+            diff_time = current_time-self._prev_time
+
+            # file eta
+            f_eta_seconds = 0
+            if previous_done_files:
+                # we assume an additional network request per slower pending file
+                # operation to estimate an upper bound (e.g. AB, AcoustID)
+                f_eta_seconds = (diff_time / previous_done_files) * (pending_files + pending_requests)
+
+            # network eta
+            n_eta_seconds = 0
+            if previous_done_requests:
+                # we assume an additional network request per pending file
+                # to estimate an upper bound (e.g. fetch cover, lookup, scan)
+                n_eta_seconds = (diff_time / previous_done_requests) * (pending_requests + pending_files)
+
+            # general eta (biased to whatever takes longer)
+            eta_seconds = max(n_eta_seconds, f_eta_seconds)
+
+            # estimate progress
+            progress = diff_time / (diff_time + eta_seconds)
+
+            if self._eta_label:
+                seconds = max(eta_seconds, 0)
+
+                # calculate eta in d:h:m:s format
+                days = int(seconds*INV_SECS_IN_DAY)
+                seconds = seconds - days*SECS_IN_DAY
+                hours = int(seconds*INV_SECS_IN_HOUR)
+                seconds = seconds - hours*SECS_IN_HOUR
+                minutes = int(seconds*INV_SECS_IN_MINUTE)
+                seconds = seconds - minutes*SECS_IN_MINUTE
```

Comment:
This could be made simpler and more generic:

```python
SECS_IN_DAY = 86400
SECS_IN_HOUR = 3600
SECS_IN_MINUTE = 60


def euclidian_div(a, b):
    return (a // b, a % b)


def seconds_to_dhms(seconds):
    days, seconds = euclidian_div(seconds, SECS_IN_DAY)
    hours, seconds = euclidian_div(seconds, SECS_IN_HOUR)
    minutes, seconds = euclidian_div(seconds, SECS_IN_MINUTE)

    return (days, hours, minutes, seconds)
```

Such functions could be moved to utils and will be easier to test.

Reply:
I thought using the integer div and mod would be slower, but ended up being faster. *the more you know*